File: app/src/main/java/com/agrobridge/data/model/Lote.kt
package com.agrobridge.data.model

import androidx.compose.ui.graphics.Color
import com.agrobridge.presentation.theme.*
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.gson.annotations.SerializedName

/**
 * Modelo de Lote
 * Representa un campo/lote de producción agrícola
 * ACTUALIZADO: Incluye geolocalización y features de iOS
 */
data class Lote(
    @SerializedName("id")
    val id: String,

    @SerializedName("nombre")
    val nombre: String,

    @SerializedName("cultivo")
    val cultivo: String,

    @SerializedName("area")
    val area: Double, // En hectáreas

    @SerializedName("estado")
    val estado: LoteEstado,

    @SerializedName("productor")
    val productor: Productor,

    @SerializedName("fechaCreacion")
    val fechaCreacion: Long, // Unix timestamp

    // ============================================================================
    // GEOLOCALIZACIÓN (replicado de iOS)
    // ============================================================================

    /**
     * Polígono del campo (lista de coordenadas GPS)
     * Si es null, no se puede mostrar en mapa
     */
    @SerializedName("coordenadas")
    val coordenadas: List<Coordenada>? = null,

    /**
     * Centro del campo (para mostrar marker en mapa)
     * Si es null, se calcula automáticamente del polígono
     */
    @SerializedName("centroCampo")
    val centroCampo: Coordenada? = null,

    // ============================================================================
    // PROPIEDADES ADICIONALES
    // ============================================================================

    @SerializedName("ubicacion")
    val ubicacion: String? = null,

    @SerializedName("bloqueId")
    val bloqueId: String? = null,

    @SerializedName("bloqueNombre")
    val bloqueNombre: String? = null,

    @SerializedName("metadata")
    val metadata: Any? = null
) {

    // ============================================================================
    // COMPUTED PROPERTIES (replicados de iOS)
    // ============================================================================

    /**
     * Color del estado en el mapa
     * Replica extension LoteEstado.mapColor de iOS
     */
    val mapColor: Color
        get() = when (estado) {
            LoteEstado.ACTIVO -> StatusActive
            LoteEstado.EN_COSECHA -> StatusHarvest
            LoteEstado.COSECHADO -> StatusCertified
            LoteEstado.EN_PREPARACION -> StatusPreparation
            LoteEstado.INACTIVO -> StatusInactive
        }

    /**
     * Emoji del cultivo
     * Replica extension String.emoji de iOS
     */
    val cultivoEmoji: String
        get() = when (cultivo.lowercase()) {
            "aguacate" -> "🥑"
            "tomate" -> "🍅"
            "maíz", "maiz" -> "🌽"
            "fresa" -> "🍓"
            "café", "cafe" -> "☕"
            "cacao" -> "🍫"
            "papa" -> "🥔"
            "chile" -> "🌶️"
            "frijol" -> "🫘"
            else -> "🌱"
        }

    /**
     * CameraPosition para Google Maps (centrada en este lote)
     * Replica var region de iOS
     */
    val cameraPosition: CameraPosition?
        get() {
            val centro = centroCampo ?: calcularCentro()
            return centro?.let {
                CameraPosition.Builder()
                    .target(it.toLatLng())
                    .zoom(16f) // Equivalente a latitudeDelta 0.01 de iOS
                    .build()
            }
        }

    /**
     * Verificar si el lote tiene datos GPS válidos
     */
    val hasValidGPS: Boolean
        get() = coordenadas != null && coordenadas.size >= 3

    /**
     * Área del polígono calculada usando fórmula de Shoelace con corrección geodética
     * Usa Radio de la Tierra para convertir a metros cuadrados, luego a hectáreas
     * Preciso para cualquier latitud (no solo en el ecuador)
     */
    val areaCalculada: Double?
        get() {
            // Safe navigation: return null if no valid GPS
            val coords = coordenadas ?: return null
            if (coords.size < 3) return null

            // Shoelace formula con proyección Cartesiana basada en latitud
            var areaM2 = 0.0
            for (i in coords.indices) {
                val j = (i + 1) % coords.size
                val lat1 = Math.toRadians(coords[i].latitud)
                val lon1 = Math.toRadians(coords[i].longitud)
                val lat2 = Math.toRadians(coords[j].latitud)
                val lon2 = Math.toRadians(coords[j].longitud)

                // Convertir a coordenadas Cartesianas aproximadas usando radio de la Tierra
                // Ajustar longitud por cos(latitud promedio) para precisión en latitudes altas
                val avgLat = (lat1 + lat2) / SHOELACE_DIVISOR
                val x1 = lon1 * Math.cos(avgLat) * EARTH_RADIUS_METERS
                val y1 = lat1 * EARTH_RADIUS_METERS
                val x2 = lon2 * Math.cos(avgLat) * EARTH_RADIUS_METERS
                val y2 = lat2 * EARTH_RADIUS_METERS

                // Aplicar fórmula de Shoelace
                areaM2 += (x1 * y2) - (x2 * y1)
            }

            // Resultado en m², dividir entre 10,000 para convertir a hectáreas
            val hectareas = Math.abs(areaM2) / SHOELACE_DIVISOR / SQ_METERS_PER_HECTARE
            return if (hectareas > 0) hectareas else null
        }

    // ============================================================================
    // MÉTODOS DE GEOLOCALIZACIÓN
    // ============================================================================

    /**
     * Calcular centro geométrico del polígono
     * Si no hay centroCampo definido, lo calcula
     */
    private fun calcularCentro(): Coordenada? {
        if (coordenadas == null || coordenadas.isEmpty()) return null

        val latPromedio = coordenadas.map { it.latitud }.average()
        val lonPromedio = coordenadas.map { it.longitud }.average()

        return Coordenada(latPromedio, lonPromedio)
    }

    /**
     * Verificar si un punto está dentro del polígono (ray casting algorithm)
     * Implementa algoritmo de ray casting seguro con tolerancia epsilon para
     * precisión de punto flotante y manejo de casos especiales (bordes verticales)
     *
     * BUG FIX HIGH-3: Previene división por cero en bordes verticales
     */
    fun contienePunto(punto: Coordenada): Boolean {
        // Safe null check
        val coords = coordenadas ?: return false
        if (coords.size < 3) return false

        var inside = false
        val epsilon = 1e-10  // Tolerancia para comparaciones de punto flotante
        var j = coords.size - 1

        for (i in coords.indices) {
            val xi = coords[i].longitud
            val yi = coords[i].latitud
            val xj = coords[j].longitud
            val yj = coords[j].latitud
            val px = punto.longitud
            val py = punto.latitud

            // Ray casting: verificar si el rayo horizontal desde punto intersecta el edge
            if ((yi > py) != (yj > py)) {
                // Edge cruza la línea horizontal del punto
                val deltaX = xj - xi
                if (Math.abs(deltaX) > epsilon) {
                    // Evitar división por cero con verificación epsilon
                    val slope = (xj - xi) / (yj - yi)
                    val xIntersect = xi + slope * (py - yi)

                    if (px < xIntersect) {
                        inside = !inside
                    }
                }
                // Si deltaX ~= 0 (borde casi vertical), ignorar para evitar inestabilidad numérica
            }
            j = i
        }

        return inside
    }

    companion object {
        // FIXED: LOW-039 - Extract geographic constants from area calculation
        /** Earth's radius in meters (WGS84 sphere approximation) */
        private const val EARTH_RADIUS_METERS = 6371000.0

        /** Shoelace formula divisor for area calculation */
        private const val SHOELACE_DIVISOR = 2.0

        /** Conversion factor from square meters to hectares */
        private const val SQ_METERS_PER_HECTARE = 10000.0

        /**
         * Datos mock para testing (replicado de iOS mockLotes)
         */
        fun mockLotes(): List<Lote> = listOf(
            Lote(
                id = "lote-001",
                nombre = "Campo Norte",
                cultivo = "Aguacate",
                area = 12.5,
                estado = LoteEstado.ACTIVO,
                productor = Productor.mock(),
                fechaCreacion = System.currentTimeMillis(),
                coordenadas = listOf(
                    Coordenada(19.432608, -99.133209),
                    Coordenada(19.433608, -99.133209),
                    Coordenada(19.433608, -99.134209),
                    Coordenada(19.432608, -99.134209)
                ),
                centroCampo = Coordenada(19.433108, -99.133709)
            ),
            Lote(
                id = "lote-002",
                nombre = "Campo Sur",
                cultivo = "Tomate",
                area = 8.3,
                estado = LoteEstado.EN_COSECHA,
                productor = Productor.mock(),
                fechaCreacion = System.currentTimeMillis(),
                coordenadas = listOf(
                    Coordenada(19.431608, -99.133209),
                    Coordenada(19.432608, -99.133209),
                    Coordenada(19.432608, -99.134209),
                    Coordenada(19.431608, -99.134209)
                ),
                centroCampo = Coordenada(19.432108, -99.133709)
            ),
            Lote(
                id = "lote-003",
                nombre = "Campo Este",
                cultivo = "Maíz",
                area = 15.7,
                estado = LoteEstado.EN_PREPARACION,
                productor = Productor.mock(),
                fechaCreacion = System.currentTimeMillis(),
                coordenadas = listOf(
                    Coordenada(19.433608, -99.132209),
                    Coordenada(19.434608, -99.132209),
                    Coordenada(19.434608, -99.133209),
                    Coordenada(19.433608, -99.133209)
                ),
                centroCampo = Coordenada(19.434108, -99.132709)
            )
        )
    }
}

/**
 * Enum de estados de lote
 * Debe coincidir con el backend
 */
enum class LoteEstado {
    @SerializedName("activo")
    ACTIVO,

    @SerializedName("inactivo")
    INACTIVO,

    @SerializedName("en_cosecha")
    EN_COSECHA,

    @SerializedName("cosechado")
    COSECHADO,

    @SerializedName("en_preparacion")
    EN_PREPARACION;

    /**
     * Display name para UI
     */
    val displayName: String
        get() = when (this) {
            ACTIVO -> "Activo"
            INACTIVO -> "Inactivo"
            EN_COSECHA -> "En Cosecha"
            COSECHADO -> "Cosechado"
            EN_PREPARACION -> "En Preparación"
        }
}

File: ios/AgroBridge/AgroBridge/Models/Lote.swift
import Foundation
import CoreLocation
import MapKit
import SwiftUI

// MARK: - Lote Model
/// Modelo que representa un lote de producción agrícola
struct Lote: Codable, Identifiable {
    let id: String
    let nombre: String
    let ubicacion: String
    let tipoCultivo: String
    let areaHectareas: Double?
    let fechaCreacion: Date
    let estado: LoteEstado
    let productorId: String
    let bloqueId: String? // ID del bloque asociado
    let bloqueNombre: String? // Nombre del bloque asociado
    let metadata: LoteMetadata?

    // MARK: - Geolocalización (MapKit)
    /// Coordenadas del polígono que representa el campo
    let coordenadas: [Coordenada]?
    /// Centro del campo para mostrar pin en el mapa
    let centroCampo: Coordenada?

    // MARK: - Computed Properties para MapKit
    /// Región del mapa centrada en este lote
    var region: MKCoordinateRegion? {
        guard let centro = centroCampo else { return nil }
        return MKCoordinateRegion(
            center: CLLocationCoordinate2D(
                latitude: centro.latitud,
                longitude: centro.longitud
            ),
            span: MKCoordinateSpan(
                latitudeDelta: 0.01,
                longitudeDelta: 0.01
            )
        )
    }

    /// Polígono de MapKit para dibujar el campo
    var polygon: MKPolygon? {
        guard let coords = coordenadas, !coords.isEmpty else { return nil }
        let locations = coords.map { coord in
            CLLocationCoordinate2D(
                latitude: coord.latitud,
                longitude: coord.longitud
            )
        }
        return MKPolygon(coordinates: locations, count: locations.count)
    }

    enum CodingKeys: String, CodingKey {
        case id
        case nombre
        case ubicacion
        case tipoCultivo = "tipo_cultivo"
        case areaHectareas = "area_hectareas"
        case fechaCreacion = "fecha_creacion"
        case estado
        case productorId = "productor_id"
        case bloqueId = "bloque_id"
        case bloqueNombre = "bloque_nombre"
        case metadata
        case coordenadas
        case centroCampo = "centro_campo"
    }
}

// MARK: - LoteEstado
enum LoteEstado: String, Codable, CaseIterable {
    case activo = "activo"
    case inactivo = "inactivo"
    case enCosecha = "en_cosecha"
    case cosechado = "cosechado"
    case enPreparacion = "en_preparacion"

    var displayName: String {
        switch self {
        case .activo: return "Activo"
        case .inactivo: return "Inactivo"
        case .enCosecha: return "En Cosecha"
        case .cosechado: return "Cosechado"
        case .enPreparacion: return "En Preparación"
        }
    }

    var color: String {
        switch self {
        case .activo: return "green"
        case .inactivo: return "gray"
        case .enCosecha: return "orange"
        case .cosechado: return "blue"
        case .enPreparacion: return "yellow"
        }
    }
}

// MARK: - LoteEstado Extensions para MapKit
extension LoteEstado {
    /// Color para mostrar en el mapa
    var mapColor: Color {
        switch self {
        case .activo:
            return Color.green
        case .enCosecha:
            return Color.orange
        case .cosechado:
            return Color.blue
        case .enPreparacion:
            return Color.yellow
        case .inactivo:
            return Color.gray
        }
    }
}

// MARK: - LoteMetadata
struct LoteMetadata: Codable {
    let coordenadasGPS: CoordenadaGPS?
    let fotos: [String]? // URLs de fotos
    let notas: String?

    enum CodingKeys: String, CodingKey {
        case coordenadasGPS = "coordenadas_gps"
        case fotos
        case notas
    }
}

// MARK: - CoordenadaGPS
struct CoordenadaGPS: Codable {
    let latitud: Double
    let longitud: Double
    let altitud: Double?

    enum CodingKeys: String, CodingKey {
        case latitud
        case longitud
        case altitud
    }
}

// MARK: - CreateLoteRequest
/// Request para endpoint POST /lotes
struct CreateLoteRequest: Codable {
    let nombre: String
    let ubicacion: String
    let tipoCultivo: String
    let areaHectareas: Double?
    let productorId: String?
    let metadata: LoteMetadata?

    enum CodingKeys: String, CodingKey {
        case nombre
        case ubicacion
        case tipoCultivo = "tipo_cultivo"
        case areaHectareas = "area_hectareas"
        case productorId = "productor_id"
        case metadata
    }
}

// MARK: - LotesResponse
/// Response del endpoint GET /lotes
struct LotesResponse: Codable {
    let lotes: [Lote]
    let total: Int
    let pagina: Int?
    let totalPaginas: Int?

    enum CodingKeys: String, CodingKey {
        case lotes
        case total
        case pagina
        case totalPaginas = "total_paginas"
    }
}

// MARK: - String Extension para Emojis de Cultivos
extension String {
    /// Emoji representativo del cultivo
    var cultivoEmoji: String {
        switch self.lowercased() {
        case "aguacate": return "🥑"
        case "tomate": return "🍅"
        case "maíz", "maiz": return "🌽"
        case "fresa": return "🍓"
        case "café", "cafe": return "☕"
        case "cacao": return "🍫"
        case "papa", "patata": return "🥔"
        case "banana", "plátano", "platano": return "🍌"
        case "naranja": return "🍊"
        case "manzana": return "🍎"
        default: return "🌱"
        }
    }
}

// MARK: - Mock Data
extension Lote {
    static var mockLotes: [Lote] {
        [
            Lote(
                id: "lote-001",
                nombre: "Campo Norte",
                ubicacion: "Valle Central, México",
                tipoCultivo: "Aguacate",
                areaHectareas: 12.5,
                fechaCreacion: Date().addingTimeInterval(-86400 * 30), // Hace 30 días
                estado: .activo,
                productorId: "prod-001",
                bloqueId: "blk-001",
                bloqueNombre: "Bloque A",
                metadata: nil,
                coordenadas: [
                    Coordenada(latitud: 19.432608, longitud: -99.133209),
                    Coordenada(latitud: 19.433608, longitud: -99.133209),
                    Coordenada(latitud: 19.433608, longitud: -99.134209),
                    Coordenada(latitud: 19.432608, longitud: -99.134209)
                ],
                centroCampo: Coordenada(latitud: 19.433108, longitud: -99.133709)
            ),
            Lote(
                id: "lote-002",
                nombre: "Campo Sur",
                ubicacion: "Valle Central, México",
                tipoCultivo: "Tomate",
                areaHectareas: 8.3,
                fechaCreacion: Date().addingTimeInterval(-86400 * 60), // Hace 60 días
                estado: .enCosecha,
                productorId: "prod-001",
                bloqueId: "blk-002",
                bloqueNombre: "Bloque B",
                metadata: nil,
                coordenadas: [
                    Coordenada(latitud: 19.431608, longitud: -99.133209),
                    Coordenada(latitud: 19.432608, longitud: -99.133209),
                    Coordenada(latitud: 19.432608, longitud: -99.134209),
                    Coordenada(latitud: 19.431608, longitud: -99.134209)
                ],
                centroCampo: Coordenada(latitud: 19.432108, longitud: -99.133709)
            ),
            Lote(
                id: "lote-003",
                nombre: "Campo Este",
                ubicacion: "Valle Central, México",
                tipoCultivo: "Maíz",
                areaHectareas: 15.7,
                fechaCreacion: Date().addingTimeInterval(-86400 * 15), // Hace 15 días
                estado: .enPreparacion,
                productorId: "prod-002",
                bloqueId: "blk-003",
                bloqueNombre: "Bloque C",
                metadata: nil,
                coordenadas: [
                    Coordenada(latitud: 19.430608, longitud: -99.133209),
                    Coordenada(latitud: 19.431608, longitud: -99.133209),
                    Coordenada(latitud: 19.431608, longitud: -99.134209),
                    Coordenada(latitud: 19.430608, longitud: -99.134209)
                ],
                centroCampo: Coordenada(latitud: 19.431108, longitud: -99.133709)
            )
        ]
    }
}
